src/adapter/src/catalog/migrate.rs

```diff
@@ -152,6 +181,163 @@ fn csr_url_path_rewrite(stmt: &mut mz_sql::ast::Statement<Raw>) {
+        name,
+        connection:
+            CreateSourceConnection::Postgres {
+                connection,
+                options,
+            },
+        ..
+    }) = stmt
+    {
+        // Find the option containing the serialized details.
+        let details_idx = options
+            .iter()
+            .position(|PgConfigOption { name, .. }| name == &PgConfigOptionName::Details)
+            .expect("corrupt catalog");
+
+        // Examine the current details
+        let details = match &options[details_idx].value {
+            Some(mz_sql::ast::WithOptionValue::Value(mz_sql::ast::Value::String(details))) => {
+                details
+            }
+            _ => unreachable!("corrupt catalog"),
+        };
+
+        let details = hex::decode(details).expect("valid catalog");
+        let details =
+            ProtoPostgresSourcePublicationDetails::decode(&*details).expect("valid catalog");
+        let mut publication_details =
+            PostgresSourcePublicationDetails::from_proto(details).expect("valid catalog");
+
+        if publication_details
+            .tables
+            .iter()
+            .all(|t| t.columns.iter().all(|c| c.col_num != 0))
+        {
+            // If every column is not 0, then no need for this migration.
+            return;
+        }
+
+        // Get details to connect to the upspsource PG instance, which we need
+        // to get the schema details.
+        let scx = StatementContext::new(None, &*catalog);
+        let connection = {
+            let item = scx.resolve_item(connection.clone()).expect("valid catalog");
+            match item.connection().expect("valid catalog") {
+                mz_storage_client::types::connections::Connection::Postgres(connection) => {
+                    connection
+                }
+                _ => unreachable!("corrupt catalog"),
+            }
+        };
+
+        let publication = options
+            .iter()
+            .find(|o| matches!(o.name, PgConfigOptionName::Publication))
+            .expect("valid catalog")
+            .value
+            .as_ref()
+            .expect("valid catalog");
+
+        let publication = match publication {
+            mz_sql::ast::WithOptionValue::Value(mz_sql::ast::Value::String(publication)) => {
+                publication
+            }
+            _ => unreachable!("corrupt catalog"),
+        };
+
+        // verify that we can connect upstream and snapshot publication metadata
+        let config = connection
+            .config(&*connection_context.secrets_reader)
+            .await
+            .expect("valid config");
+
+        // Get the current publication tables from the upstream PG source.
+        let current_publication_tables =
+            match mz_postgres_util::publication_info(&config, publication, None).await {
+                Ok(v) => v,
+                Err(_) => {
+                    warn!(
```

Comment:
Should this situation the status of the source to be updated in mz_source_statuses? Right now it will be just a `WARN` message in the logs, right?

What would be an "external dependency" in this case? Should the reason for the error be part of the message rather than consumed and discarded via `_` ?

Reply:
Nevermind, the error does end up in `mz_source_statuses` , so all good.

Reply:
The error in the mz_source_statuses is `publication "mz_source" does not exist` . The longer warning is only in the log, so not actionable by users.

Reply:
Yes, I think that behavior is right. I don't want the migration to prevent Materialize from booting and whatever the issue is will get surfaced elsewhere. The warnings here are just for our debugging purposes.